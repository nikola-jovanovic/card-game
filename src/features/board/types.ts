import { TypeOf } from 'io-ts'

import { Card as CardC } from './entities/Card'

export type Card = TypeOf<typeof CardC>
export type Score = number

export enum Names {
  Me = 'Me',
  Milisav = 'Milisav',
  Mileva = 'Mileva',
  Djura = 'Djura',
}

export interface Player {
  name: Names
  score: Score
  cards: Card[]
  active: boolean
}

export type Pile = {
  [key in Names]?: Card
}