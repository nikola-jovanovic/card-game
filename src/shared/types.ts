export type Action<P, M = void> = {
  type: string
  payload: P
  meta: M
}

