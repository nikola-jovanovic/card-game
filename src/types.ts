import { TypeOf } from 'io-ts'
import { Card as CardC } from './entities/Card'

export type Action<P, M = void> = {
  type: string
  payload: P
  meta: M
}

export type Card = TypeOf<typeof CardC>
