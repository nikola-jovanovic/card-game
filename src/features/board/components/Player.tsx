import React, { memo } from 'react'
import styled from 'styled-components'

import Title from '../../../shared/components/Title'
import { Player as PlayerT } from '../types'
import Card from './Card'

export type Props = PlayerT & {
  variant: 'user' | 'cpu'
  onPlay?: any
}

const CardWrapper = styled.div<{ padding: boolean }>`
  padding-left: ${(props) => (props.padding ? 50 : 0)}px;

  & > img {
    margin-left: -50px;
  }
`

const Name = styled(Title) <{ color: string }>`
  color: ${props => props.color};
  margin: 0 0 5px;
`

const Score = styled.p`
  margin: 0 0 10px
`

const Player = ({
  name,
  score,
  cards,
  onPlay,
  variant,
}: Props): JSX.Element => {
  return (
    <div>
      <Name as="h2" color={variant === 'cpu' ? 'white' : 'yellow'}>{name}</Name>
      <Score>Score: {score}</Score>
      <CardWrapper padding={variant === 'cpu'}>
        {cards.map((card) => (
          <Card key={card.code} {...card} facedown={variant === 'cpu'} onClick={onPlay} />
        ))}
      </CardWrapper>
    </div>
  )
}

export default memo(Player)
