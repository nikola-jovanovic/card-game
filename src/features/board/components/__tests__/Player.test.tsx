import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'

import Player, { Props } from '../Player'
import { Names } from '../../types'

describe('Player', () => {
  let props: Props

  beforeAll(() => {
    props = {
      name: Names.Me,
      score: 50,
      cards: [{
        code: 'JH',
        image: 'image-url',
        value: 5
      }],
      variant: 'user',
      onPlay: jest.fn(),
      active: true
    }
  })

  it('user', () => {
    const { container } = render(<Player {...props} />)

    const button = screen.getByRole('button', { name: /jh/i })

    expect(button).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /jh/i })).toBeInTheDocument()
    expect(screen.getByRole('heading', { name: /me/i, level: 2 })).toBeInTheDocument()

    expect(container).toMatchSnapshot()

    fireEvent.click(button)

    expect(props.onPlay).toHaveBeenCalledTimes(1)
    expect(props.onPlay).toHaveBeenCalledWith(props.cards[0])
  })

  it('cpu', () => {
    const { container } = render(<Player {...props} variant='cpu' onPlay={undefined} />)

    expect(screen.getByRole('img', { name: /facedown/i })).toBeInTheDocument()
    expect(screen.getByRole('heading', { name: /me/i, level: 2 })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })
})