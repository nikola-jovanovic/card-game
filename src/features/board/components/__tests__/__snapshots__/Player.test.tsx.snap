// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player cpu 1`] = `
.c0 {
  color: #fff;
  margin: 0;
  margin-bottom: 10px;
}

.c4 {
  max-width: 80px;
}

.c3 {
  padding-left: 50px;
}

.c3 > img {
  margin-left: -50px;
}

.c1 {
  color: white;
  margin: 0 0 5px;
}

.c2 {
  margin: 0 0 10px;
}

<div>
  <div>
    <h2
      class="c0 c1"
      color="white"
    >
      Me
    </h2>
    <p
      class="c2"
    >
      Score: 
      50
    </p>
    <div
      class="c3"
    >
      <img
        aria-label="facedown card"
        class="c4"
        src="https://dcassetcdn.com/design_img/3930521/154340/26190370/kfmqknarypcs90a4zrbf3ppavp_thumbnail.png"
      />
    </div>
  </div>
</div>
`;

exports[`Player user 1`] = `
.c0 {
  color: #fff;
  margin: 0;
  margin-bottom: 10px;
}

.c5 {
  max-width: 80px;
}

.c4 {
  border: none;
  padding: 0;
  margin-right: 4px;
  background-color: transparent;
  cursor: pointer;
}

.c3 {
  padding-left: 0px;
}

.c3 > img {
  margin-left: -50px;
}

.c1 {
  color: yellow;
  margin: 0 0 5px;
}

.c2 {
  margin: 0 0 10px;
}

<div>
  <div>
    <h2
      class="c0 c1"
      color="yellow"
    >
      Me
    </h2>
    <p
      class="c2"
    >
      Score: 
      50
    </p>
    <div
      class="c3"
    >
      <button
        aria-label="JH"
        class="c4"
      >
        <img
          aria-label="JH"
          class="c5"
          src="image-url"
        />
      </button>
    </div>
  </div>
</div>
`;
